text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,461 @@
+
+Here's how `OnceCell` can be used to implement lazy-initialized global data:
+
+```rust
+use std::{sync::{Mutex, OnceCell}, collections::HashMap};
+
+fn global_data() -> &'static Mutex<HashMap<i32, String>> {
+    static INSTANCE: OnceCell<Mutex<HashMap<i32, String>>> = OnceCell::new();
+    INSTANCE.get_or_init(|| {
+        let mut m = HashMap::new();
+        m.insert(13, "Spica".to_string());
+        m.insert(74, "Hoyten".to_string());
+        Mutex::new(m)
+    })
+}
+```
+
+Here's how `OnceCell` can be used to implement a lazy field:
+
+```rust
+use std::{fs, io, path::PathBuf, cell::OnceCell};
+
+struct Ctx {
+    config_path: PathBuf,
+    config: OnceCell<String>,
+}
+
+impl Ctx {
+    pub fn get_config(&self) -> Result<&str, io::Error> {
+        let cfg = self.config.get_or_try_init(|| {
+            fs::read_to_string(&self.config_path)
+        })?;
+        Ok(cfg.as_str())
+    }
+}
+```
+
+We also provide a more convenient but less powerful `Lazy<T, F>` wrapper around `OnceCell<T>`, which allows to specify the initializing closure at creation time:
+
+```rust
+pub struct Lazy<T, F = fn() -> T> { ... }
+
+impl<T, F> Lazy<T, F> {
+    /// Creates a new lazy value with the given initializing function.
+    pub const fn new(init: F) -> Lazy<T, F>;
+}
+
+impl<T, F: FnOnce() -> T> Lazy<T, F> {
+    /// Forces the evaluation of this lazy value and returns a reference to
+    /// the result.
+    ///
+    /// This is equivalent to the `Deref` impl, but is explicit.
+    pub fn force(this: &Lazy<T, F>) -> &T;
+}
+
+impl<T, F: FnOnce() -> T> Deref for Lazy<T, F> {
+    type Target = T;
+
+    fn deref(&self) -> &T;
+}
+```
+
+`Lazy` directly replaces `lazy_static!`:
+
+```rust
+use std::{sync::{Mutex, Lazy}, collections::HashMap};
+
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>> = Lazy::new(|| {
+    let mut m = HashMap::new();
+    m.insert(13, "Spica".to_string());
+    m.insert(74, "Hoyten".to_string());
+    Mutex::new(m)
+});
+```
+
+Moreover, once `#[thread_local]` attribute is stable, `Lazy` might supplant `std::thread_local!` as well:
+
+```rust
+use std::cell::{RefCell, Lazy};
+
+#[thread_local]
+pub static FOO: Lazy<RefCell<u32>> = Lazy::new(|| RefCell::new(1));
+```
+
+Unlike `lazy_static!`, `Lazy` can be used used for locals:
+
+```rust
+use std::cell::Lazy;
+
+fn main() {
+    let ctx = vec![1, 2, 3];
+    let thunk = Lazy::new(|| {
+        ctx.iter().sum::<i32>()
+    });
+    assert_eq!(*thunk, 6);
+}
+```
+
+However, `#[thread_local]` attribute is pretty far from stabilization at the moment, and due to the required special handling of destructors, it's unclear if just using `cell::Lazy` will work out.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The proposed API is directly copied from [`once_cell`] crate.
+
+Altogether, this RFC proposes to add four types:
+
+* `std::cell::OnceCell`, `std::cell::Lazy`
+* `std::sync::OnceCell`, `std::sync::Lazy`
```

Comment:
I see that we have [`sync::Once`](https://doc.rust-lang.org/std/sync/struct.Once.html) and [`cell::Cell`](https://doc.rust-lang.org/std/cell/struct.Cell.html), but this may be confusing. Just a thought.

Reply:
This is kinda like the combination of both, so the naming may be good (modulo `Send`/`Sync` concerns). But it would be better if the `!Sync` version had a different name from the `Sync` version.

Reply:
Perhaps acknowledging `OnceCell` as a WORM (write once, read many) might make `WormCell` be a better name? <del>Assuming that the implementation isn't too sluggish.</del>

Reply:
I also would really like to see there be a different name for these, rather than relying on the fact that they lie in two separate modules. Maybe just prefixing the sync implementations with `Sync` is enough, i.e. `SyncLazy` and `Lazy`.

Reply:
I would be against WORM terminology, as it seems rather obscure: I don't think I see it often enough to not consult wikipedia. A more appropriate term would probably be `SingleAssignmentCell`, but that's a mouthful.

Ultimately, I doubt that there's a perfect solution for cell/sync naming, so the libs team would have to just decide on some particular naming scheme. I am personally ok with `Sync` prefix. The disadvantage of `SyncX` scheme is that it makes more common case longer to spell, and is a tautology (`sync::SyncOnceCell`).

I personally like the proposed `cell::OnceCell / sync::OnceCell` variant the most:

* if you want extra, you can qualify the type with `cell::` or `sync::` on the call site, instead of importing it
* due to how awesome Rust is, it's impossible to use `cell::OnceCell` instead of `sync::OnceCell`, and the mistake in other direction is not really a mistake: just a very slight perf loss
* in practice, I think it would be rare for a module to use `cell` and `sync` versions simultaneously.

Reply:
I think I like the way it's organized in the `once_cell` crate with explicit `sync::*` and `unsync::*` modules.

The part that feels the most strange to me about this currently proposal is that a `OnceCell` lives in both `cell` (makes sense, since it's a type of cell), and `sync` (makes sense, given that `Once` is also defined here). Something's conflated here, though this could well be outside the scope of this RFC.

Reply:
Yeah, another possibility is to introduce something like lazy and lazy::sync modules. As it stands, std::sync today feels a bit like, well, kitchen sync, it has a bit of everything.

I can imagine an alternative world, where std::sync is only for synchronization primitives, atomic and mpsc are top-level modules and Arc lives in rc. In this world, having an std::lazy module would definitely be better.

With current stdlib structure, std::sync::OnceCell blends better with the background.

Reply:
Also, everything in `std::cell` is `!Sync` so it sort of maps to crates.io’s `once_cell::unsync`.

Reply:
The libs team has discussed moving stuff out of std::sync into more appropriate modules (e.g. std::mutex::Mutex). I think ideally we wouldn't throw more stuff in there.